File: src/main/scala/Geometry.scala
package main.scala

sealed trait Shape

case class Point(x: Double, y: Double) {}

case class Square(topLeft: Point, side: Double) extends Shape {}

case class Rectangle(topLeft: Point, height: Double, width: Double) extends Shape {}

case class Circle(center: Point, radius: Double) extends Shape {}

class Geometry {
    val PI = 3.141592653589793

    def area(shape: Shape): Double = {
        shape match {
            case Square(_, s) => s * s
            case Rectangle(_, h, w) => h * w
            case Circle(_, r) => PI * r * r
        }
    }
}

File: src/main/kotlin/Geometry.kt
package main.kotlin

sealed interface Shape

data class Point(val x: Double, val y: Double) {}

data class Square(val topLeft: Point, val side: Double): Shape {}

data class Rectangle(val topLeft: Point, val height: Double, val width: Double): Shape {}

data class Circle(val center: Point, val radius: Double): Shape {}

class Geometry {
    val PI = 3.141592653589793

    fun area(shape: Shape): Double {
        return when(shape) {
            is Square -> shape.side * shape.side
            is Rectangle -> shape.height * shape.width
            is Circle -> PI * shape.radius * shape.radius
        }
    }
}
